cpp/src/arrow/dataset/filter.cc

```diff
@@ -758,43 +797,107 @@ Result<std::shared_ptr<DataType>> FieldExpression::Validate(const Schema& schema
   return null();
 }
 
-struct AccumulateFieldsVisitor {
-  void operator()(const FieldExpression& expr) { fields.push_back(expr.name()); }
+Result<std::shared_ptr<Expression>> InsertImplicitCasts(const Expression& expr,
+                                                        const Schema& schema) {
+  struct {
+    struct Validated {
+      std::shared_ptr<Expression> expr;
+      std::shared_ptr<DataType> type;
+    };
+
+    Result<Validated> Validate(const Expression& expr) {
```

Comment:
Perhaps this is the kind of case where returning a Status and passing two out-parameters is simpler ;-)